### ISSUE TITLE
compilation fails with ffmpeg 7.0

### ISSUE THREAD

**Issue by VVD** [OpenShot/libopenshot#953]
FreeBSD 14.0 amd64, ffmpeg 7.0.

Part of build log:
```
FAILED: src/CMakeFiles/openshot.dir/FFmpegReader.cpp.o 
/usr/bin/c++ -DHAVE_ISFINITE=1 -DQT_CORE_LIB -DQT_GUI_LIB -DQT_NO_DEBUG -DQT_SVG_LIB -DQT_WIDGETS_LIB -DUSE_BABL=1 -DUSE_HW_ACCEL=1 -DUSE_IMAGEMAGICK=1 -DUSE_LEGACY_TRACKER=1 -DUSE_OPENCV=1 -Dopenshot_EXPORTS -I/wrkdirs/usr/ports/multimedia/libopenshot/work-py39/.build/src/openshot_autogen/include -I/wrkdirs/usr/ports/multimedia/libopenshot/work-py39/libopenshot-0.3.2/src -I/wrkdirs/usr/ports/multimedia/libopenshot/work-py39/.build/src -isystem /usr/local/include/libopenshot-audio -isystem /usr/local/include/ImageMagick-7 -isystem /usr/local/include -isystem /usr/local/include/qt5 -isystem /usr/local/include/qt5/QtCore -isystem /usr/local/lib/qt5/mkspecs/freebsd-clang -isystem /usr/local/include/qt5/QtGui -isystem /usr/local/include/qt5/QtWidgets -isystem /usr/local/include/qt5/QtSvg -isystem /usr/local/include/babl-0.1 -isystem /usr/local/include/opencv4 -O2 -pipe  -fstack-protector-strong -fno-strict-aliasing   -std=c++17 -O2 -pipe  -fstack-protector-strong -fno-strict-aliasing   -std=c++17  -DNDEBUG -std=c++17 -fPIC -DMAGICKCORE_HDRI_ENABLE=1 -DMAGICKCORE_QUANTUM_DEPTH=16 -DMAGICKCORE_CHANNEL_MASK_DEPTH=32 -fPIC -fopenmp=libomp -pthread -MD -MT src/CMakeFiles/openshot.dir/FFmpegReader.cpp.o -MF src/CMakeFiles/openshot.dir/FFmpegReader.cpp.o.d -o src/CMakeFiles/openshot.dir/FFmpegReader.cpp.o -c /wrkdirs/usr/ports/multimedia/libopenshot/work-py39/libopenshot-0.3.2/src/FFmpegReader.cpp
/wrkdirs/usr/ports/multimedia/libopenshot/work-py39/libopenshot-0.3.2/src/FFmpegReader.cpp:676:51: error: no member named 'channel_layout' in 'AVCodecParameters'
        if (AV_GET_CODEC_ATTRIBUTES(aStream, aCodecCtx)->channel_layout == 0)
            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~  ^
/wrkdirs/usr/ports/multimedia/libopenshot/work-py39/libopenshot-0.3.2/src/FFmpegReader.cpp:677:48: error: no member named 'channel_layout' in 'AVCodecParameters'
                AV_GET_CODEC_ATTRIBUTES(aStream, aCodecCtx)->channel_layout = av_get_default_channel_layout(AV_GET_CODEC_ATTRIBUTES(aStream, aCodecCtx)->channels);
                ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~  ^
/wrkdirs/usr/ports/multimedia/libopenshot/work-py39/libopenshot-0.3.2/src/FFmpegReader.cpp:677:140: error: no member named 'channels' in 'AVCodecParameters'
                AV_GET_CODEC_ATTRIBUTES(aStream, aCodecCtx)->channel_layout = av_get_default_channel_layout(AV_GET_CODEC_ATTRIBUTES(aStream, aCodecCtx)->channels);
                                                                                                            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~  ^
/wrkdirs/usr/ports/multimedia/libopenshot/work-py39/libopenshot-0.3.2/src/FFmpegReader.cpp:688:63: error: no member named 'channels' in 'AVCodecParameters'
        info.channels = AV_GET_CODEC_ATTRIBUTES(aStream, aCodecCtx)->channels;
                        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~  ^
/wrkdirs/usr/ports/multimedia/libopenshot/work-py39/libopenshot-0.3.2/src/FFmpegReader.cpp:689:85: error: no member named 'channel_layout' in 'AVCodecParameters'
        info.channel_layout = (ChannelLayout) AV_GET_CODEC_ATTRIBUTES(aStream, aCodecCtx)->channel_layout;
                                              ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~  ^
/wrkdirs/usr/ports/multimedia/libopenshot/work-py39/libopenshot-0.3.2/src/FFmpegReader.cpp:1599:60: error: no member named 'channels' in 'AVCodecParameters'
                                                                                           AV_GET_CODEC_ATTRIBUTES(aStream, aCodecCtx)->channels,
                                                                                           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~  ^
/wrkdirs/usr/ports/multimedia/libopenshot/work-py39/libopenshot-0.3.2/src/FFmpegReader.cpp:1604:91: error: no member named 'channels' in 'AVCodecParameters'
                packet_samples = audio_frame->nb_samples * AV_GET_CODEC_ATTRIBUTES(aStream, aCodecCtx)->channels;
                                                           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~  ^
/wrkdirs/usr/ports/multimedia/libopenshot/work-py39/libopenshot-0.3.2/src/FFmpegReader.cpp:1667:88: error: no member named 'channel_layout' in 'AVCodecParameters'
        av_opt_set_int(avr, "in_channel_layout", AV_GET_CODEC_ATTRIBUTES(aStream, aCodecCtx)->channel_layout, 0);
                                                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~  ^
/wrkdirs/usr/ports/multimedia/libopenshot/work-py39/libopenshot-0.3.2/src/FFmpegReader.cpp:1668:89: error: no member named 'channel_layout' in 'AVCodecParameters'
        av_opt_set_int(avr, "out_channel_layout", AV_GET_CODEC_ATTRIBUTES(aStream, aCodecCtx)->channel_layout, 0);
                                                  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~  ^
9 errors generated.
ninja: build stopped: subcommand failed.
*** Error code 1
```
Full build log is here: https://pkg-status.freebsd.org/beefy22/data/140amd64-default-foo/2024-05-04_07h10m46s/logs/errors/py39-libopenshot-0.3.2_3.log

Maybe this can help: https://code.videolan.org/videolan/vlc/-/merge_requests/5304/diffs
`AV_GET_CODEC_ATTRIBUTES(aStream, aCodecCtx)->channels` => `AV_GET_CODEC_ATTRIBUTES(aStream, aCodecCtx)->ch_layout.nb_channels`

**Comment by eclipseo** [comment]
Oh I was working on this for Fedora Linux: here's my PR https://github.com/OpenShot/libopenshot/pull/954

**Comment by VVD** [comment]
> Oh I was working on this for Fedora Linux: here's my PR #954

Thanks.
But this patch need rework to apply to release version.

Or apply this commit from repo before your patch: https://github.com/OpenShot/libopenshot/commit/9aa37136513e06c940776f5dc7636b349ceade27

**Comment by eclipseo** [comment]
> > Oh I was working on this for Fedora Linux: here's my PR #954
> 
> Thanks. But this patch need rework to apply to release version.
> 
> Or apply this commit from repo before your patch: [9aa3713](https://github.com/OpenShot/libopenshot/commit/9aa37136513e06c940776f5dc7636b349ceade27)

What do you mean, my patch is on top of the develop branch, so it already includes this commit in the history?

If you want the version to apply on top of the current stable version: https://src.fedoraproject.org/fork/eclipseo/rpms/libopenshot/blob/07481b00585dfa14a2a3aa8986892a9f88002338/f/0001-Add-compatibility-with-FFMPEG-7.0.patch

**Comment by jonoomph** [comment]
I've left some feedback on https://github.com/OpenShot/libopenshot/pull/954. Hopefully we can get this merged version soon, and get it into a new production release of OpenShot.

**Comment by VVD** [comment]
> What do you mean, my patch is on top of the develop branch, so it already includes this commit in the history?
> 
> If you want the version to apply on top of the current stable version: https://src.fedoraproject.org/fork/eclipseo/rpms/libopenshot/blob/07481b00585dfa14a2a3aa8986892a9f88002338/f/0001-Add-compatibility-with-FFMPEG-7.0.patch

Yes, I mean on top of the last release.
But if new release is soon then we can wait.

**Comment by antonio-rojas** [comment]
This fixes build, but the are many test failures when compiled against ffmpeg 7, such as
```
 92/188 Test  #92: FrameMapper:resample_audio_48000_to_41000 ......................***Failed    0.84 sec
[SWR @ 0x5582654a9a80] Input channel layout "" is invalid or unsupported.
[SWR @ 0x5582654a9a80] Context has not been initialized
Filters: "resample_audio_48000_to_41000"
```

**Comment by lfos** [comment]
@eclipseo Are you able to take a look at the test failures? I created a separate issue for tracking, #954.

**Comment by VVD** [comment]
Is this fixed in OpenShot 3.2.0?

**Comment by antonio-rojas** [comment]
> Is this fixed in OpenShot 3.2.0?

If you mean libopenshot 0.3.3, no.

**Comment by VVD** [comment]
> > Is this fixed in OpenShot 3.2.0?
> 
> If you mean libopenshot 0.3.3, no.

And in 0.4.0?
